Svc/SeqDispatcher/SeqDispatcher.cpp

```diff
@@ -0,0 +1,210 @@
+
+SeqDispatcher ::~SeqDispatcher() {}
+
+void SeqDispatcher::init(NATIVE_INT_TYPE queueDepth,
+                         NATIVE_INT_TYPE instance) {
+  SeqDispatcherComponentBase::init(queueDepth, instance);
+  for (FwIndexType i = 0; i < SeqDispatcherSequencerPorts; i++) {
+    // assert if any of the cmd sequencer connections aren't
+    // connected. otherwise we might think we sent out
+    // a sequence to be run but it doesn't get run
+    if (!this->isConnected_seqRunOut_OutputPort(i)) {
+      FW_ASSERT(0, i);
+    }
+  }
+}
+
+FwIndexType SeqDispatcher::getNextAvailableSequencerIdx() {
+  for (FwIndexType i = 0; i < SeqDispatcherSequencerPorts; i++) {
+    if (this->m_entryTable[i].state == SeqDispatcher_CmdSequencerState::AVAILABLE) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+bool SeqDispatcher::runSequence(FwIndexType sequencerIdx,
+                                const Fw::StringBase& fileName,
+                                Fw::Wait block) {
+  FW_ASSERT(sequencerIdx >= 0 && sequencerIdx < SeqDispatcherSequencerPorts,
+            sequencerIdx);
+
+  // if the seqRunOut port at this idx isn't connected
+  if (!this->isConnected_seqRunOut_OutputPort(sequencerIdx)) {
+    this->log_WARNING_HI_INVALID_SEQUENCER(static_cast<U16>(sequencerIdx));
+    return false;
+  }
+
+  // if the sequencer is not available to run a new sequence
+  if (this->m_entryTable[sequencerIdx].state !=
+      SeqDispatcher_CmdSequencerState::AVAILABLE) {
+    this->log_WARNING_HI_INVALID_SEQUENCER(static_cast<U16>(sequencerIdx));
+    return false;
+  }
+
+  // ok, good to use this sequencer
+  if (block == Fw::Wait::NO_WAIT) {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK;
+  } else {
+    this->m_entryTable[sequencerIdx].state =
+        SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK;
+  }
+
+  this->m_sequencersAvailable--;
+  this->tlmWrite_sequencersAvailable(this->m_sequencersAvailable);
+  this->m_entryTable[sequencerIdx].sequenceRunning = fileName;
+
+  this->m_dispatchedCount++;
+  this->tlmWrite_dispatchedCount(this->m_dispatchedCount);
+  this->seqRunOut_out(sequencerIdx,
+                      this->m_entryTable[sequencerIdx].sequenceRunning);
+
+  return true;
+}
+
+void SeqDispatcher::seqStartIn_handler(
+    NATIVE_INT_TYPE portNum,                                //!< The port number
+    const Fw::StringBase& fileName //!< The sequence file name
+) {
+  FW_ASSERT(portNum >= 0 && portNum < SeqDispatcherSequencerPorts, portNum);
+  if (this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_BLOCK ||
+      this->m_entryTable[portNum].state ==
+          SeqDispatcher_CmdSequencerState::RUNNING_SEQUENCE_NO_BLOCK) {
+    // we were aware of this sequencer running a sequence
+    if (this->m_entryTable[portNum].sequenceRunning != fileName) {
+      // uh oh. m_sequencesRunning is wrong
```

Comment:
I think this needs to produce a WARNING_HI event.  This is because:

1. You've said to run a sequence
2. Something else in the system ran a different sequence

I agree there is not automatic resolution, but the requested sequence is likely being rejected by the sequence engine.

A more complicated behavior might be to retry dispatch to another member of the pool....but we'd need to think through the implications.

Reply:
Agreed, I will add a WARNING_HI